datastore/datastore.go

```diff
@@ -0,0 +1,322 @@
+}
+
+// InventoryData is the data type for inventory data produced by a plugin data
+// source and emitted to the agent's inventory data store
+type InventoryData map[string]interface{}
+
+// MetricData is the data type for events produced by a plugin data source and
+// emitted to the agent's metrics data store
+type MetricData map[string]interface{}
+
+// EventData is the data type for single shot events
+type EventData map[string]interface{}
+
+//StackdriverMetric represents fields for stackdriver returns
+type StackdriverMetric struct {
+	TimeSeries []struct {
+		Metric struct {
+			Labels struct {
+				ApiMethod    string `json:"api_method"`
+				ResponseCode string `json:"response_code"`
+			} `json:"labels"`
+			Type string `json:"type"`
+		} `json:"metric"`
+		MetricKind string `json:"metricKind"`
+		Points     []struct {
+			Interval struct {
+				EndTime   time.Time `json:"endTime"`
+				StartTime time.Time `json:"startTime"`
+			} `json:"interval"`
+			Value struct {
+				Int64Value int64 `json:"int64Value,string"`
+			} `json:"value"`
+		} `json:"points"`
+		Resource struct {
+			Labels struct {
+				ModuleID  string `json:"module_id"`
+				ProjectID string `json:"project_id"`
+				VersionID string `json:"version_id"`
+			} `json:"labels"`
+			Type string `json:"type"`
+		} `json:"resource"`
+		ValueType string `json:"valueType"`
+	} `json:"timeSeries"`
+}
+
+func Run(log *logrus.Logger, prettyPrint bool, version string) {
+	var data = PluginData{
+		Name:            NAME,
+		ProtocolVersion: ProtocolVersion,
+		PluginVersion:   version,
+		Inventory:       make(map[string]InventoryData),
+		Metrics:         make([]MetricData, 0),
+		Events:          make([]EventData, 0),
+	}
+
+	c, projectId, err := NewClient()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	//add query metrics
+	kinds, err := DatastoreStatKindQueryResult(c.Dsc)
+	if err != nil {
+		log.Fatal(err)
+	}
+	result := DatastoreData(kinds, projectId)
+
+	for _, metricResult := range result {
+		data.Metrics = append(data.Metrics, metricResult)
+	}
+
+	//add stackdriver metrics
+	for _, metric := range stackdriverEndpoints {
+		resp, err := StackdriverResp(projectId, metric)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		result, err := StackdriverData(resp)
+		if err != nil {
+			log.Fatal(err)
+		}
+		for _, metricResult := range result {
+			data.Metrics = append(data.Metrics, metricResult)
+		}
+	}
+
+	fatalIfErr(log, helpers.OutputJSON(data, prettyPrint))
+}
+
+//NewClient() creates a client for datastore, it primarily exists for testing purposes
+func NewClient() (Client, string, error) {
+	dsClient, projectId, err := ConnectDatastore(base64Config)
+	if err != nil {
+		return Client{}, "", err
+	}
+
+	return Client{
+		dsClient,
+	}, projectId, nil
+}
+
+//DatastoreStatKindQueryResult performs a Query against the datastore and return values using __Stat_Kind__
+func DatastoreStatKindQueryResult(ds DatastoreClient) ([]DatastoreKind, error) {
```

Comment:
Rather than taking DatastoreClient as the first argument I think it makes sense to make this a method of the client type, ie `func (c *Client) KindStats() ([]DatastoreKind, error)`